### PR TITLE
Add SDPA support for T5 Style Models

### LINKED ISSUE #28005: Open to contribution: adding `torch.nn.functional.scaled_dot_product_attention` support for more architectures
### Feature request

In [`Transformers 4.36`](https://github.com/huggingface/transformers/releases/tag/v4.36.0), we started adding native support of [torch.nn.functional.scaled_dot_product_attention](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention.html) (SDPA), enabled by default in Transformers: https://huggingface.co/docs/transformers/perf_infer_gpu_one#flashattention-and-memory-efficient-attention-through-pytorchs-scaleddotproductattention

SDPA allows to dispatch to memory-efficient attention, flash attention on supported GPUs (currently NVIDIA-only), and even on [Intel CPUs](https://pytorch.org/blog/new-features-for-ai/#flash-attention-based-scaled-dot-product-algorithm-for-cpu).

For the record, here's a benchmark on some currently supported models:

**[Training benchmark](https://gist.github.com/fxmarty/7e75cc3942d6974e4849093ebea0a331), run on A100-SXM4-80GB.**

| Model     | Batch size | Sequence length | Time per batch (`"eager"`, s) | Time per batch (`"sdpa"`, s) | **Speedup** | Peak memory (`"eager"`, MB) | Peak memory (`"sdpa"`, MB) | **Memory savings**    |
|-----------|------------|-----------------|-------------------------------|------------------------------|-------------|-----------------------------|----------------------------|-----------------------|
| llama2 7b | 4          | 1024            | 1.065                         | 0.90                         | **19.4%**   | 73878.28                    | 45977.81                   | **60.7%**             |
| llama2 7b | 4          | 2048            | OOM                           | 1.87                         | /           | OOM                         | 78394.58                   | **SDPA does not OOM** |
| llama2 7b | 1          | 2048            | 0.64                          | 0.48                         | **32.0%**   | 55557.01                    | 29795.63                   | **86.4%**             |
| llama2 7b | 1          | 3072            | OOM                           | 0.75                         | /           | OOM                         | 37916.08                   | **SDPA does not OOM** |
| llama2 7b | 1          | 4096            | OOM                           | 1.03                         | /           | OOM                         | 46028.14                   | **SDPA does not OOM** |
| llama2 7b | 2          | 4096            | OOM                           | 2.05                         | /           | OOM                         | 78428.14                   | **SDPA does not OOM** |

**[Inference benchmark](https://gist.github.com/fxmarty/5113e4304fbdd38c9c3702ce44683f6a), run on A100-SXM4-80GB.**

| Model            | Batch size | Prompt length | Num new tokens | Per token latency `"eager"` (ms) | Per token latency `"sdpa"` (ms) | **Speedup** |
|------------------|------------|---------------|----------------|----------------------------------|---------------------------------|-------------|
| llama2 13b       | 1          | 1024          | 1 (prefill)    | 178.66                           | 159.36                          | **12.11%**  |
| llama2 13b       | 1          | 100           | 100            | 40.35                            | 37.62                           | **7.28%**   |
| llama2 13b       | 8          | 100           | 100            | 40.55                            | 38.06                           | **6.53%**   |
| Whisper v3 large | 1          | /             | 62             | 20.05                            | 18.90                           | **6.10%**   |
| Whisper v3 large | 8          | /             | 77             | 25.42                            | 24.77                           | **2.59%**   |
| Whisper v3 large | 16         | /             | 77             | 28.51                            | 26.32                           | **8.34%**   |

Previously, we had a partial support of SDPA in [Optimum BetterTransformer](https://huggingface.co/docs/optimum/bettertransformer/overview) but we are now looking to slowly deprecate it in favor of upstream support of SDPA directly in Transformers.

Here are the architectures for which support has been requested:
- [ ] Codegen (https://github.com/huggingface/optimum/issues/1050)
- [ ] LLAVA (https://github.com/huggingface/optimum/issues/1592)
- [ ] Marian (https://github.com/huggingface/optimum/issues/1142)
- [x] Mistral (https://github.com/huggingface/optimum/issues/1553)
- [ ] LongT5 (https://github.com/huggingface/optimum/issues/1506)
- [ ] ViT (https://github.com/huggingface/optimum/issues/1553)

The integration could take inspiration from https://github.com/huggingface/optimum/blob/main/optimum/bettertransformer/models/decoder_models.py & https://github.com/huggingface/optimum/blob/main/optimum/bettertransformer/models/attention.py

### Motivation

Faster training & inference, lower memory requirement

### Your contribution

I may work on some at some point, but contributions are most welcome.

You should refer to https://github.com/huggingface/transformers/pull/26572 to add the support of SDPA for a model, roughly following these steps:
* Create a `XxxSdpaAttention` class inheriting from `XxxAttention` and implement the attention logic using SDPA
* Use `_prepare_4d_causal_attention_mask_for_sdpa` instead of `_prepare_4d_causal_attention_mask` for SDPA
* Use `_prepare_4d_attention_mask_for_sdpa` instead of `_prepare_4d_attention_mask` for SDPA
* Add `_supports_sdpa = True` to `XxxPreTrainedModel`
* Add `"sdpa"` key to `XXX_ATTENTION_CLASSES` in the model modeling file

### PR DESCRIPTION
**Pull request by abdulfatir**
# What does this PR do?

Adds `torch`'s `scaled_dot_product_attention` support for the T5 model.

Part of #28005

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @fxmarty @sayakpaul

### CODE REVIEW

**Comment by abdulfatir** [comment]
I ran tests for T5 locally (`pytest tests/models/t5/test_modeling_t5.py`). Most of the tests pass except the following:

```
FAILED tests/models/t5/test_modeling_t5.py::T5ModelTest::test_disk_offload_bin - AssertionError: ValueError not raised
FAILED tests/models/t5/test_modeling_t5.py::T5ModelTest::test_prompt_lookup_decoding_matches_greedy_search - RuntimeError: The size of tensor a (2) must match the size of tensor b (3) at non-singleton dimension 3
FAILED tests/models/t5/test_modeling_t5.py::T5ModelTest::test_torch_fx - AssertionError: Couldn't trace module: __bool__ should return bool, returned Tensor
FAILED tests/models/t5/test_modeling_t5.py::T5ModelTest::test_torch_fx_output_loss - AssertionError: Couldn't trace module: __bool__ should return bool, returned Tensor
```

I am not exactly sure about the failure reason for the first two. It looks like the last two may be failing due to the `if self.is_decoder` control flow (not sure). Is the fix here making separate encoder and decoder classes? I would like to get some suggestions on this before I go ahead and fix other T5-like models. 

Another question: If `sdpa` is supported by a model, would it automatically be incorporated into the current tests or should I also write some unit tests to ensure that the `sdpa` and `eager` versions match?

File: src/transformers/models/t5/modeling_t5.py [hunk 1]
```diff
@@ -1028,11 +1183,27 @@ def forward(
...
+        if self._use_sdpa and head_mask is None and not output_attentions:
+            extended_attention_mask = (
+                _prepare_4d_causal_attention_mask_for_sdpa(
+                    attention_mask,
+                    input_shape,
+                    inputs_embeds,
+                    past_key_values_length,
+                )
+                if self.is_decoder
+                else _prepare_4d_attention_mask_for_sdpa(attention_mask, inputs_embeds.dtype)
```
Comment (abdulfatir) [inline src/transformers/models/t5/modeling_t5.py:1199]: Is this causing tracing test failures?

**Comment by sayakpaul** [comment]
Thanks @abdulfatir. Did you get to compare the speed-up provided by SDPA? If not, no worries.

**Comment by abdulfatir** [comment]
@ArthurZucker @fxmarty any thoughts on this? How should I proceed?

File: src/transformers/models/t5/modeling_t5.py [hunk 2]
```diff
@@ -575,10 +581,157 @@ def project(hidden_states, proj_layer, key_value_states, past_key_value):
...
+                elif past_key_value.shape[2] != key_value_states.shape[1]:
+                    # checking that the `sequence_length` of the `past_key_value` is the same as
+                    # the provided `key_value_states` to support prefix tuning
+                    # cross-attn
+                    # (batch_size, n_heads, seq_length, dim_per_head)
+                    hidden_states = shape(proj_layer(key_value_states))
+                else:
+                    # cross-attn
+                    hidden_states = past_key_value
+            return hidden_states
```
Comment (fxmarty) [inline src/transformers/models/t5/modeling_t5.py:665]: Personal taste, but I would rather redefine those as staticmethods in `T5Attention`, avoiding the duplicated code.

**Review by fxmarty — COMMENTED** [review]
Thank you @abdullaholuk! For the failing tests you mention, could you try running them on the main branch and check their status there? It may be they are failing there as well (if they are slow tests).

Could you try as well: `RUN_SLOW=True pytest tests/models/t5 -k "test_eager_matches_sdpa_inference" -s -vvvvv`

File: src/transformers/models/t5/modeling_t5.py (hunk 2 above)
Comment (abdulfatir) [reply src/transformers/models/t5/modeling_t5.py:665]: Thanks! I will do that.

**Comment by abdulfatir** [comment]
> Thank you @abdullaholuk! For the failing tests you mention, could you try running them on the main branch and check their status there? It may be they are failing there as well (if they are slow tests).
> 
> Could you try as well: `RUN_SLOW=True pytest tests/models/t5 -k "test_eager_matches_sdpa_inference" -s -vvvvv`

@fxmarty Thanks! These tests passed (will check on a `bf16` supported device later):

```
tests/models/t5/test_modeling_t5.py::T5ModelTest::test_eager_matches_sdpa_inference_0_float16 <- tests/test_modeling_common.py PASSED
tests/models/t5/test_modeling_t5.py::T5ModelTest::test_eager_matches_sdpa_inference_1_bfloat16 <- tests/test_modeling_common.py SKIPPED (bfloat16 not supported on cuda (on the
specific device currently used, e.g. Nvidia T4 GPU))
tests/models/t5/test_modeling_t5.py::T5ModelTest::test_eager_matches_sdpa_inference_2_float32 <- tests/test_modeling_common.py PASSED
tests/models/t5/test_modeling_t5.py::T5EncoderOnlyModelTest::test_eager_matches_sdpa_inference_0_float16 <- tests/test_modeling_common.py PASSED
tests/models/t5/test_modeling_t5.py::T5EncoderOnlyModelTest::test_eager_matches_sdpa_inference_1_bfloat16 <- tests/test_modeling_common.py SKIPPED (bfloat16 not supported on cuda (on
the specific device currently used, e.g. Nvidia T4 GPU))
tests/models/t5/test_modeling_t5.py::T5EncoderOnlyModelTest::test_eager_matches_sdpa_inference_2_float32 <- tests/test_modeling_common.py PASSED
```

I would like to get your thoughts on the tracing test failures. Have you seen this before and do you know why these could be failing? If not, I can take a deeper look to figure out what's wrong.
Also, do you have any thoughts on the `stride` issue I mentioned in [a comment above](https://github.com/huggingface/transformers/pull/30375#issuecomment-2074706941).

File: src/transformers/models/t5/modeling_t5.py (hunk 2 above)
Comment (abdulfatir) [reply src/transformers/models/t5/modeling_t5.py]: Addressed.

**Comment by abdulfatir** [comment]
@fxmarty @ArthurZucker 

I addressed the comments. The `test_prompt_lookup_decoding_matches_greedy_search` test is failing and I am not completely sure why. Do you have any insights? 

Once we fix this and everything else looks okay, I will fix the copies of this model.

File: src/transformers/models/t5/modeling_t5.py [hunk 3]
```diff
@@ -572,10 +590,137 @@ def project(hidden_states, proj_layer, key_value_states, past_key_value):
...
+        if self.pruned_heads:
+            mask = torch.ones(position_bias.shape[1])
+            mask[list(self.pruned_heads)] = 0
+            position_bias_masked = position_bias[:, mask.bool()]
+        else:
+            position_bias_masked = position_bias
+
+        # spda kernels require tensors to have stride=1 in the last dimension
+        # .contiguous() does not behave correctly for tensors with singleton dimensions
+        # .clone(memory_format=torch.contiguous_format) is a workaround
```
Comment (fxmarty) [inline src/transformers/models/t5/modeling_t5.py:691]: You could like to https://github.com/pytorch/pytorch/issues/127523 here

**Comment by fxmarty** [comment]
@abdulfatir is there still the issue with
```
FAILED tests/models/t5/test_modeling_t5.py::T5ModelTest::test_torch_fx - AssertionError: Couldn't trace module: __bool__ should return bool, returned Tensor
FAILED tests/models/t5/test_modeling_t5.py::T5ModelTest::test_torch_fx_output_loss - AssertionError: Couldn't trace module: __bool__ should return bool, returned Tensor
```
?

File: src/transformers/models/t5/modeling_t5.py (hunk 3 above)
Comment (abdulfatir) [reply src/transformers/models/t5/modeling_t5.py]: Added.

**Comment by abdulfatir** [comment]
@fxmarty I tried to debug `test_prompt_lookup_decoding_matches_greedy_search` and the following is happening:

At some point [`to_4d`](https://github.com/huggingface/transformers/blob/eed9ed679878ada2f6d2eefccdbda368cabc88b1/src/transformers/modeling_attn_mask_utils.py#L97) is called with `attention_mask_2d.shape = [1, 2]`, `query_length=2`, `key_value_length=1`. As a result, [this piece of code](https://github.com/huggingface/transformers/blob/eed9ed679878ada2f6d2eefccdbda368cabc88b1/src/transformers/modeling_attn_mask_utils.py#L114-L127) creates a `causal_4d_mask` of shape `[1, 1, 2, 3]` but then tries to fill it with [`expanded_attn_mask`](https://github.com/huggingface/transformers/blob/eed9ed679878ada2f6d2eefccdbda368cabc88b1/src/transformers/modeling_attn_mask_utils.py#L137) which has shape `[1, 1, 2, 2]` causing the failure. Honestly, I found it a bit difficult to parse what's the expected behavior. Can you give some insight @fxmarty? I would like to close this as soon as possible.

**Comment by abdulfatir** [comment]
@fxmarty @ArthurZucker I fixed the issues. Tried fixing the repo consistency to the best of my ability. Can you please take a look?

**Review by ArthurZucker — COMMENTED** [review]
You solved the fix copies, but pop2piano seems to not be exportable to onnx anymore!

**Comment by abdulfatir** [comment]
@ArthurZucker I changed it to use `clone` instead of `copy_`  but now the test breaks with:

> FAILED tests/models/pop2piano/test_modeling_pop2piano.py::Pop2PianoModelTest::test_export_to_onnx - torch.onnx.errors.UnsupportedOperatorError: Exporting the operator 'aten::scaled_dot_product_attention' to ONNX opset version 9 is not supported. Support for this operator was added in version 14, try exporting with this version.

But I believe this should also be an issue for other sdpa calls, right? Why is only this test breaking?

**Comment by abdulfatir** [comment]
Tried a higher `opset_version` as suggested by the error but something is still failing with SDPA + ONNX. This maybe relevant: https://github.com/pytorch/pytorch/issues/96944 cc @fxmarty

**Comment by fxmarty** [comment]
@abdulfatir you can remove this test. Transformers ONNX export has been deprecated for >1 year now.

**Comment by abdulfatir** [comment]
Thanks @fxmarty! I skipped the test. Can you please also review the PR (and the run the remaining workflows)?

**Review by fxmarty — APPROVED** [review]
LGTM as long as the slow tests pass & some real models have been validated

**Comment by abdulfatir** [comment]
@fxmarty for slow tests do we need a label?

**Comment by fxmarty** [comment]
@abdulfatir I don't think we can run them from github actions. You just need to run `RUN_SLOW=1 pytest tests/models/t5 -s -vvvvv` and check that no more tests fail than on main branch (some might already be failing).

File: src/transformers/models/longt5/modeling_longt5.py [hunk 4]
```diff
@@ -539,7 +549,9 @@ def project(hidden_states, proj_layer, key_value_states, past_key_value):
         if layer_head_mask is not None:
             attn_weights = attn_weights * layer_head_mask
 
-        attn_output = unshape(torch.matmul(attn_weights, value_states))  # (batch_size, seq_length, dim)
+        attn_output = self._unshape(
+            torch.matmul(attn_weights, value_states), batch_size
+        )  # (batch_size, seq_length, dim)
```
Comment (amyeroberts) [inline src/transformers/models/longt5/modeling_longt5.py:554]:
```suggestion
        # (batch_size, seq_length, dim)
        attn_output = self._unshape(torch.matmul(attn_weights, value_states), batch_size)
```

File: src/transformers/models/mt5/modeling_mt5.py [hunk 5]
```diff
@@ -348,50 +389,25 @@ def forward(
...
-                else:
-                    # cross-attn
-                    hidden_states = past_key_value
-            return hidden_states
-
         # get query states
-        query_states = shape(self.q(hidden_states))  # (batch_size, n_heads, seq_length, dim_per_head)
+        query_states = self._shape(
+            self.q(hidden_states), batch_size
+        )  # (batch_size, n_heads, seq_length, dim_per_head)
```
Comment (amyeroberts) [inline src/transformers/models/mt5/modeling_mt5.py:395]:
```suggestion
        # (batch_size, n_heads, seq_length, dim_per_head)
        query_states = self._shape(self.q(hidden_states), batch_size)
```

File: src/transformers/models/mt5/modeling_mt5.py [hunk 6]
```diff
@@ -436,7 +452,9 @@ def project(hidden_states, proj_layer, key_value_states, past_key_value):
         if layer_head_mask is not None:
             attn_weights = attn_weights * layer_head_mask
 
-        attn_output = unshape(torch.matmul(attn_weights, value_states))  # (batch_size, seq_length, dim)
+        attn_output = self._unshape(
+            torch.matmul(attn_weights, value_states), batch_size
+        )  # (batch_size, seq_length, dim)
```
Comment (amyeroberts) [inline src/transformers/models/mt5/modeling_mt5.py:457]:
```suggestion
        # (batch_size, seq_length, dim)
        attn_output = self._unshape(torch.matmul(attn_weights, value_states), batch_size)
```

File: src/transformers/models/pop2piano/modeling_pop2piano.py [hunk 7]
```diff
@@ -379,50 +414,25 @@ def forward(
...
-                else:
-                    # cross-attn
-                    hidden_states = past_key_value
-            return hidden_states
-
         # get query states
-        query_states = shape(self.q(hidden_states))  # (batch_size, n_heads, seq_length, dim_per_head)
+        query_states = self._shape(
+            self.q(hidden_states), batch_size
+        )  # (batch_size, n_heads, seq_length, dim_per_head)
```
Comment (amyeroberts) [inline src/transformers/models/pop2piano/modeling_pop2piano.py:420]:
```suggestion
        # (batch_size, n_heads, seq_length, dim_per_head)
        query_states = self._shape(self.q(hidden_states), batch_size)
```

File: src/transformers/models/pop2piano/modeling_pop2piano.py [hunk 8]
```diff
@@ -467,7 +477,9 @@ def project(hidden_states, proj_layer, key_value_states, past_key_value):
         if layer_head_mask is not None:
             attn_weights = attn_weights * layer_head_mask
 
-        attn_output = unshape(torch.matmul(attn_weights, value_states))  # (batch_size, seq_length, dim)
+        attn_output = self._unshape(
+            torch.matmul(attn_weights, value_states), batch_size
+        )  # (batch_size, seq_length, dim)
```
Comment (amyeroberts) [inline src/transformers/models/pop2piano/modeling_pop2piano.py:482]:
```suggestion
        # (batch_size, seq_length, dim)
        attn_output = self._unshape(torch.matmul(attn_weights, value_states), batch_size)
```

File: src/transformers/models/switch_transformers/modeling_switch_transformers.py [hunk 9]
```diff
@@ -569,7 +579,9 @@ def project(hidden_states, proj_layer, key_value_states, past_key_value):
         if layer_head_mask is not None:
             attn_weights = attn_weights * layer_head_mask
 
-        attn_output = unshape(torch.matmul(attn_weights, value_states))  # (batch_size, seq_length, dim)
+        attn_output = self._unshape(
+            torch.matmul(attn_weights, value_states), batch_size
+        )  # (batch_size, seq_length, dim)
```
Comment (amyeroberts) [inline src/transformers/models/switch_transformers/modeling_switch_transformers.py:584]:
```suggestion
        # (batch_size, seq_length, dim)
        attn_output = self._unshape(torch.matmul(attn_weights, value_states), batch_size)
```

File: src/transformers/models/switch_transformers/modeling_switch_transformers.py [hunk 10]
```diff
@@ -481,50 +516,25 @@ def forward(
...
-                else:
-                    # cross-attn
-                    hidden_states = past_key_value
-            return hidden_states
-
         # get query states
-        query_states = shape(self.q(hidden_states))  # (batch_size, n_heads, seq_length, dim_per_head)
+        query_states = self._shape(
+            self.q(hidden_states), batch_size
+        )  # (batch_size, n_heads, seq_length, dim_per_head)
```
Comment (amyeroberts) [inline src/transformers/models/switch_transformers/modeling_switch_transformers.py:522]:
```suggestion
        # (batch_size, n_heads, seq_length, dim_per_head)
        query_states = self._shape(self.q(hidden_states), batch_size)
```

File: src/transformers/models/t5/modeling_t5.py [hunk 11]
```diff
@@ -473,50 +514,25 @@ def forward(
...
-                else:
-                    # cross-attn
-                    hidden_states = past_key_value
-            return hidden_states
-
         # get query states
-        query_states = shape(self.q(hidden_states))  # (batch_size, n_heads, seq_length, dim_per_head)
+        query_states = self._shape(
+            self.q(hidden_states), batch_size
+        )  # (batch_size, n_heads, seq_length, dim_per_head)
```
Comment (amyeroberts) [inline src/transformers/models/t5/modeling_t5.py:520]:
```suggestion
        # (batch_size, n_heads, seq_length, dim_per_head)
        query_states = self._shape(self.q(hidden_states), batch_size)
```

File: src/transformers/models/t5/modeling_t5.py [hunk 12]
```diff
@@ -561,7 +577,9 @@ def project(hidden_states, proj_layer, key_value_states, past_key_value):
         if layer_head_mask is not None:
             attn_weights = attn_weights * layer_head_mask
 
-        attn_output = unshape(torch.matmul(attn_weights, value_states))  # (batch_size, seq_length, dim)
+        attn_output = self._unshape(
+            torch.matmul(attn_weights, value_states), batch_size
+        )  # (batch_size, seq_length, dim)
```
Comment (amyeroberts) [inline src/transformers/models/t5/modeling_t5.py:582]:
```suggestion
        # (batch_size, seq_length, dim)
        attn_output = self._unshape(torch.matmul(attn_weights, value_states), batch_size)
```

File: src/transformers/models/udop/modeling_udop.py [hunk 13]
```diff
@@ -732,50 +767,25 @@ def forward(
...
-                else:
-                    # cross-attn
-                    hidden_states = past_key_value
-            return hidden_states
-
         # get query states
-        query_states = shape(self.q(hidden_states))  # (batch_size, n_heads, seq_length, dim_per_head)
+        query_states = self._shape(
+            self.q(hidden_states), batch_size
+        )  # (batch_size, n_heads, seq_length, dim_per_head)
```
Comment (amyeroberts) [inline src/transformers/models/udop/modeling_udop.py:773]:
```suggestion
        # (batch_size, n_heads, seq_length, dim_per_head)
        query_states = self._shape(self.q(hidden_states), batch_size)
```

File: src/transformers/models/udop/modeling_udop.py [hunk 14]
```diff
@@ -820,7 +830,9 @@ def project(hidden_states, proj_layer, key_value_states, past_key_value):
         if layer_head_mask is not None:
             attn_weights = attn_weights * layer_head_mask
 
-        attn_output = unshape(torch.matmul(attn_weights, value_states))  # (batch_size, seq_length, dim)
+        attn_output = self._unshape(
+            torch.matmul(attn_weights, value_states), batch_size
+        )  # (batch_size, seq_length, dim)
```
Comment (amyeroberts) [inline src/transformers/models/udop/modeling_udop.py:835]:
```suggestion
        # (batch_size, seq_length, dim)
        attn_output = self._unshape(torch.matmul(attn_weights, value_states), batch_size)
```

File: src/transformers/models/longt5/modeling_longt5.py [hunk 15]
```diff
@@ -1007,6 +1019,7 @@ def unshape(states):
 
 # Copied from transformers.models.t5.modeling_t5.T5LayerSelfAttention with T5->LongT5
 class LongT5LayerSelfAttention(nn.Module):
```
Comment (amyeroberts) [inline src/transformers/models/longt5/modeling_longt5.py:1021]: What's the reason for not adding support for LongT5?

**Review by amyeroberts — COMMENTED** [review]
Thanks for all the work adding this! 

cc @ylacombe regarding pop2piano

All of the models which have this added should have information added to their modeling page, on how to use and speedups e.g. [like here for Mistral](https://huggingface.co/docs/transformers/v4.41.3/en/model_doc/mistral#speeding-up-mistral-by-using-flash-attention)

Are SDPA tests run for all of these models now? 

The slow integration tests will need to be run for all the models with SDPA added. They can be triggered by pushing a commit with the message: `[run_slow] pop2piano,mt5,t5`. Another HF member or I will need to approve the workflow for it to run

**Comment by ylacombe** [comment]
Hey @amyeroberts, thanks for pinging, I'll take a closer look to pop2piano once your commentaries have been addressed if that's okay with you!

A small comment from my side @abdulfatir, have you been able [to test speed-ups again](https://github.com/huggingface/transformers/pull/30375#issuecomment-2074706941) now that you found the reasons for the apparent lack of speed-ups ?

Also, to properly measure inference speed-ups, you should also modify a bit the script you've used to:
1. set a fixed number of tokens to generate
2. use [torch.cuda.Event](https://pytorch.org/docs/stable/generated/torch.cuda.Event.html) to properly measure time spent in inference 

I haven't tested the following script but it should work with little to any modifs!



```python
import argparse
import time

import torch

from transformers import (
    GenerationConfig,
    T5ForConditionalGeneration,
    T5TokenizerFast,
)
from transformers import set_seed


def bytes_to_giga_bytes(bytes):
    return bytes / 1024 / 1024 / 1024


def load_model(args):
    model = T5ForConditionalGeneration.from_pretrained(
        "google/flan-t5-large",
        attn_implementation=args.attn_implementation,
        torch_dtype=getattr(torch, args.torch_dtype),
    ).to("cuda")
    tokenizer = T5TokenizerFast.from_pretrained("google/flan-t5-large")
    return model, tokenizer


def get_inputs(args, tokenizer):
    inputs = tokenizer(
        ["Pedro Pedro Pedro 🦝🙌" * 10] * args.batch_size,
        padding=True,
        return_tensors="pt",
    )
    inputs = {k: v.to("cuda") for k, v in inputs.items()}
    return inputs

def measure_latency_and_memory_use(model, device, inputs, generation_config, nb_loops):

  # define Events that measure start and end of the generate pass
  start_event = torch.cuda.Event(enable_timing=True)
  end_event = torch.cuda.Event(enable_timing=True)

  # reset cuda memory stats and empty cache
  torch.cuda.reset_peak_memory_stats(device)
  torch.cuda.empty_cache()
  torch.cuda.synchronize()

  # get the start time
  start_event.record()

  # actually generate
  for _ in range(nb_loops):
        # set seed for reproducibility
        set_seed(0)
        generation = model.generate(**inputs, generation_config=generation_config)

  # get the end time
  end_event.record()
  torch.cuda.synchronize()

  # measure memory footprint and elapsed time
  max_memory = torch.cuda.max_memory_allocated(device)
  elapsed_time = start_event.elapsed_time(end_event) * 1.0e-3


  execution_time_in_s =  elapsed_time/nb_loops
  max_memory_footprint_in_GB = bytes_to_giga_bytes(max_memory)

  return execution_time_in_s, max_memory_footprint_in_GB


@torch.no_grad()
def main(args):
    model, tokenizer = load_model(args)
    inputs = get_inputs(args, tokenizer)
    generation_config = GenerationConfig(min_new_tokens=args.num_tokens, max_new_tokens=args.num_tokens)
    # warmup
    for _ in range(5):
        _ = model.generate(**inputs, generation_config=generation_config)

    avg_inference_time, max_memory_footprint_in_GB = measure_latency_and_memory_use(model, model.device, inputs, generation_config, args.num_iters)

    print(
        f"{args.attn_implementation=}, {args.torch_dtype=}, {args.batch_size=}, {args.num_iters=}, {args.num_tokens=}"
    )
    print(f"avg_inference_time: {avg_inference_time:.3f} seconds")
    print(
        f"Max memory allocated: {max_memory_footprint_in_GB} GB"
    )


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--attn_implementation", type=str, choices=["sdpa", "eager"], default="eager"
    )
    parser.add_argument(
        "--torch_dtype",
        type=str,
        choices=["float32", "bfloat16", "float16"],
        default="float32",
    )
    parser.add_argument("--batch_size", type=int, default=2)
    parser.add_argument("--num_iters", type=int, default=10)
    parser.add_argument("--num_tokens", type=int, default=20)
    args = parser.parse_args()
    main(args)
```

**Comment by abdulfatir** [comment]
Thanks @amyeroberts and @ylacombe! I will probably look into this later over the weekend.

**Comment by agemagician** [comment]
@abdulfatir any update on when this PR will be merged ?

**Comment by alvaropp** [comment]
Hi 👋🏽  any updates on this?

**Comment by ArthurZucker** [comment]
I think #34089 enabled it!

**Comment by alvaropp** [comment]
> I think #34089 enabled it!

Hi @ArthurZucker, I've tested compiling the T5 model and I do get a very modest speedup:

```python
# %%
import time

import numpy as np
import torch
from tqdm.auto import tqdm
from transformers import AutoModelForSeq2SeqLM, AutoTokenizer

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")


# %% Example prompt
tokenizer = AutoTokenizer.from_pretrained("google-t5/t5-base")

text_input = "The theory of relativity usually encompasses two interrelated physics theories by Albert Einstein: special relativity and general relativity, proposed and published in 1905 and 1915, respectively. Special relativity applies to all physical phenomena in the absence of gravity. General relativity explains the law of gravitation and its relation to the forces of nature. It applies to the cosmological and astrophysical realm, "
input_ids = tokenizer(text_input, return_tensors="pt").input_ids.to(device)

# %% Without compilation
model = AutoModelForSeq2SeqLM.from_pretrained("google-t5/t5-base").to(device)

# First pass to initialise any caches/JITs/whatever
_ = model.generate(input_ids, max_length=500)

times = []
for _ in tqdm(range(20)):
    start = time.time()
    _ = model.generate(input_ids, max_length=500)
    times.append(time.time() - start)

print(f"Times without compilation: {np.mean(times):.4f} ± {np.std(times):.4f}")

# %% With compilation
model = AutoModelForSeq2SeqLM.from_pretrained("google-t5/t5-base").to(device)
compiled_model = torch.compile(model)

# First pass to initialise any caches/JITs/whatever
_ = compiled_model.generate(input_ids, max_length=500)

times = []
for _ in tqdm(range(20)):
    start = time.time()
    _ = compiled_model.generate(input_ids, max_length=500)
    times.append(time.time() - start)

print(f"Times with compilation: {np.mean(times):.4f} ± {np.std(times):.4f}")
```

However, I've tried using:
```python
model = AutoModelForSeq2SeqLM.from_pretrained(
    "google-t5/t5-base", device_map="cuda:1", attn_implementation="SDPA"
)
model = AutoModelForSeq2SeqLM.from_pretrained(
    "google-t5/t5-base", device_map="cuda:1", attn_implementation="flash_attention_2"
)
```

but it complains that neither SDPA nor FlashAttention are implemented for the T5 model.

Any ideas? 😄

**Comment by ArthurZucker** [comment]
Ah sorry it's not implemented indeed ! 
Will handle it after #34282 !

**Comment by alvaropp** [comment]
> Ah sorry it's not implemented indeed ! Will handle it after #34282 !

Thanks, will keep an eye open!

**Comment by rubenweitzman** [comment]
Hi @ArthurZucker  wondering where the state of affair is for a t5 with sdpa/flash with working compile?

**Comment by ArthurZucker** [comment]
Now that the PR is merged, if you want you can open one for T5 or I'll do it in a bit!

**Pull request closed without merging**